### ISSUE TITLE
How the PIcontroller implemented in this package differs from that of Hairer

### ISSUE THREAD

**Issue by zhaofeng-shu33** [SciML/OrdinaryDiffEq.jl#1527]
By modulization I think the solution results of `DP5` and `DP8` produced with `OrdinaryDiffEq.jl` differs from the original fortran
code of Hairer. For example, there is [a treatment](https://github.com/pv/dopri5_interface/blob/ae2a913f12155001dbb8894575a9132117ab69ef/dopri5.f#L526) to shrink the new step when the last try is rejected in Hairer's implementation, which he did not explain in his textbook. I could not find the corresponding
line of code in this package.
```julia
sol = solve(prob, DP5(), controller=IController(), abstol=1e-6, reltol=1e-3)
```
I wonder how is this difference evaluated.

**Comment by ChrisRackauckas** [comment]
>For example, there is a treatment to shrink the new step when the last try is rejected in Hairer's implementation, which he did not explain in his textbook.

No, the treatment to shrink the time step [is here](https://github.com/pv/dopri5_interface/blob/ae2a913f12155001dbb8894575a9132117ab69ef/dopri5.f#L530), which corresponds to the [step reject controller](https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/src/integrators/controllers.jl#L157). But you are pointing out something interesting. Instead that's a stabilizer, where if the [previous step is a rejection](https://github.com/pv/dopri5_interface/blob/ae2a913f12155001dbb8894575a9132117ab69ef/dopri5.f#L531) then [the next step takes the minimum of the new step and the previous upon accept](https://github.com/pv/dopri5_interface/blob/ae2a913f12155001dbb8894575a9132117ab69ef/dopri5.f#L526). Though that's a weird way. Indeed we aren't doing that piece, but we could add it. @YingboMa thoughts?

We do have tests that things match Hairer's implementation [up to floating point error](https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/test/odeinterface/odeinterface_regression.jl), though there is a drift that can happen on some very large problems and this might explain that. It might finally be the answer to one of the oldest issues in the repo, https://github.com/SciML/OrdinaryDiffEq.jl/issues/64

**Comment by zhaofeng-shu33** [comment]
`scipy.integrate.solve_ivp` makes this feature compulsory for `DP5`,`BS3` and `DP853`.  I made [a PR](https://github.com/scipy/scipy/pull/14500) on this issue.
By the way, I could not get exactly the same solution result for `BS3` or `DP5` with IController between the Python implementation
and this Julia package, even if I used the PR mentioned above. What is the cause of this difference? Is is related with the slightly
different implementation of the local error estimation?

**Comment by zhaofeng-shu33** [comment]
I found that the accuracy defects comes from `OrdinaryDiffEq`. The critical line is [here](https://github.com/SciML/OrdinaryDiffEq.jl/blob/e3da7e00415dc55730e8e3c35438b39b13a08378/src/integrators/controllers.jl#L65). In the `IController`, the power operation is computed using float32 number. Though it could speed up the computation,
I think it produces noticeable numerical difference compared with other packages. By changing this line to use normal
power `EEst ^ expo`. I can obtain the almost the same numerical results with `scipy.integrate.solve_ivp` for `BS3` and `DP5`.

**Comment by ChrisRackauckas** [comment]
>I found that the accuracy defects
>I can obtain the almost the same numerical results with scipy.integrate.solve_ivp for BS3 and DP5.

That's not a measure of accuracy, that's a measure of same-ness. If you check the accuracy, the solvers are fine.

**Comment by ChrisRackauckas** [comment]
We could add an option to swap back the floating point pow implementation in the controllers, though using a same-accuracy default which is faster is fine. IIRC, it's a 3ulp version, so the floating point error the solver itself is much larger than the `^` error, and the `^` computation is a heuristic which is not even close to 3ulp accurate anyways. So that's what I am saying about accuracy: if you check the actual work-precision accuracy it does not effect the solver's ability to produce accurate solutions, but yes it will give a floating point difference in the low end values of the stepping behavior (which can grow over time, but washes out in terms of an actual difference)

**Comment by zhaofeng-shu33** [comment]
> > I found that the accuracy defects
> > I can obtain the almost the same numerical results with scipy.integrate.solve_ivp for BS3 and DP5.
> 
> That's not a measure of accuracy, that's a measure of same-ness. If you check the accuracy, the solvers are fine.

Yes it is the sameness.

**Comment by zhaofeng-shu33** [comment]
> 3ulp

Its indeed a good techniques of acceleration for ODE solvers using float64 type.

**Comment by zhaofeng-shu33** [comment]
> We could add an option to swap back the floating point pow implementation in the controllers, though using a same-accuracy default which is faster is fine. IIRC, it's a 3ulp version, so the floating point error the solver itself is much larger than the `^` error, and the `^` computation is a heuristic which is not even close to 3ulp accurate anyways. So that's what I am saying about accuracy: if you check the actual work-precision accuracy it does not effect the solver's ability to produce accurate solutions, but yes it will give a floating point difference in the low end values of the stepping behavior (which can grow over time, but washes out in terms of an actual difference)

Its indeed a good techniques of acceleration for ODE solvers using float64 type.
